Add fairshare metrics collection per account

ffly90 opened the pull request:
- Add new --slurm.collect-fairshare flag to enable fairshare collection
- Implement FairShareCollector using sshare command with levelfs output
- Export slurm_account_fairshare metric with account label
- Filter out empty and inf values from sshare output
- Include comprehensive unit tests for parsing logic

ffly90 committed:
Add fairshare metrics collection per account

- Add new --slurm.collect-fairshare flag to enable fairshare collection
- Implement FairShareCollector using sshare command with levelfs output
- Export slurm_account_fairshare metric with account label
- Filter out empty and inf values from sshare output
- Include comprehensive unit tests for parsing logic

abhinavDhulipala commented on exporter/fairshare.go:
```diff
@@ -0,0 +1,156 @@
+	cache        *AtomicThrottledCache[FairShareMetric]
+}
+
+func (fsf *FairShareFetcher) parseFairShare(output string) ([]FairShareMetric, error) {
```
Would it be possible to switch this to a CSV based parsing? I've done this before recently for our node cli. It can be used as an exact template for [fairshare](https://github.com/rivosinc/prometheus-slurm-exporter/blob/main/exporter/nodes.go#L156)

abhinavDhulipala commented on exporter/fairshare.go:
```diff
@@ -0,0 +1,156 @@
+package exporter
```
Unfortunately, for rivos OSS projects we need to add the following header to every file for licensing reasons. 


```suggestion
// SPDX-FileCopyrightText: 2023 Rivos Inc.
//
// SPDX-License-Identifier: Apache-2.0

package exporter
```

abhinavDhulipala commented on exporter/fairshare.go:
```diff
@@ -0,0 +1,156 @@
+func NewFairShareCollector(config *Config) *FairShareCollector {
+	return &FairShareCollector{
+		fetcher: &FairShareFetcher{
+			scraper: NewCliScraper("sshare", "-n", "-P", "-o", "account,levelfs"),
```
Let's parameterize the config to pass in a custom cmd if desired. For example, say a user has wrote their own slurm plugin that outputs compatible information or want to explicitly point to a different binary. Please add fairshare `string[]` [here](https://github.com/rivosinc/prometheus-slurm-exporter/blob/main/exporter/server.go#L25), add the default in the [constructor](https://github.com/rivosinc/prometheus-slurm-exporter/blob/main/exporter/server.go#L86), and invoke it as such:



```suggestion
            cliOpts := config.cliOpts
            
			scraper: NewCliScraper(cliOpts.newOpt...),
```

abhinavDhulipala commented on exporter/fairshare_test.go:
```diff
@@ -0,0 +1,113 @@
+package exporter
```
Ditto on the reuse-header

abhinavDhulipala commented on exporter/fairshare_test.go:
```diff
@@ -0,0 +1,113 @@
+	"testing"
+)
+
+func TestParseFairShare(t *testing.T) {
```
Let's use the `assert` and `require` pkg's here to be more expressive and inline with the rest of the pkg


```suggestion
func TestParseFairShare(t *testing.T) {
    require := require.New(t)
    assert := assert.New(t)
```

abhinavDhulipala commented on exporter/fairshare_test.go:
```diff
@@ -0,0 +1,113 @@
+	metrics, err := fetcher.parseFairShare(output)
+	if err != nil {
+		t.Fatalf("Unexpected error: %v", err)
+	}
```
```suggestion
	require.NoError(err, "failed to parse fair share error")
```

abhinavDhulipala commented on main.go:
```diff
@@ -31,6 +31,7 @@ var (
 	slurmLicEnabled      = flag.Bool("slurm.collect-licenses", false, "Collect license info from slurm")
 	slurmDiagEnabled     = flag.Bool("slurm.collect-diags", false, "Collect daemon diagnostics stats from slurm")
 	slurmSacctEnabled    = flag.Bool("slurm.collect-limits", false, "Collect account and user limits from slurm")
+	fairshareEnabled     = flag.Bool("slurm.collect-fairshare", false, "Collect fairshare info per account from slurm")
```
As mentioned before, let's add another string arg for `slurm.sshare-cli`

abhinavDhulipala reviewed (changes requested):
Howdy! Firstly, amazing job and thanks a ton for contributing to our project. The code is very good for a first time contribution. High level feedback:

  - Let's parameterize the cli invocation
  - Convert to a more robust csv parsing
  - Use the `require` and `assert` libs in testing
  - (Optional) use and expand the custom NAFloat type or another custom type to unmarshal your fairshare data.

Thanks again for your contribution. Once these changes have been incorporated, we will try it in our staging environment to ensure that the field sizes don't need to be tuned, and the parsing works in production. Then, we'll get it merged and I'll launch deploy a new minor version w/ associated pkgs.

We can expand upon this feature later with the rest of the data we can tease out from sshare and improve on this incrementally once this mvp is in:

```
sshare --helpformat
Account           Cluster           EffectvUsage      FairShare        
GrpTRESMins       GrpTRESRaw        ID                LevelFS          
NormShares        NormUsage         Partition         RawShares        
RawUsage          TRESRunMins       User  
```

abhinavDhulipala commented:
Also, is there any chance you can share a img of your panel. Bonus points if you can toss me some JSON so I can add this panel to our starter template.

abhinavDhulipala commented:
Trying it out locally, I'm getting some good data! It seems to be working well

Commit:
Add Makefile and spec-file for rpm build + systemd service

ffly90 commented:
Thank you for your detailed feedback! I will get back to you next week. Unfortunatly I will not have time to look at it prior to that.

ffly90 committed:
Refactor fairshare to use CSV-based parsing

- Replace manual string parsing with encoding/csv Reader
- Follow same pattern as AccountCsvFetcher for consistency
- Add TrimLeadingSpace to handle indented account names
- Update tests to use StringByteScraper for inline test data
- All tests passing with new CSV implementation

ffly90 committed:
Parameterize fairshare CLI command for custom overrides

- Add fairshare []string field to CliOpts
- Add SlurmFairshareOverride to CliFlags
- Add --slurm.fairshare-cli flag for custom command override
- Default to 'sshare -n -P -o Account,NormShares'
- Follow same pattern as other CLI overrides (sinfo, squeue, etc.)
- Allows users to point to custom plugins or alternative binaries

ffly90 committed:
Use require and assert for more expressive test assertions

- Add require.New(t) and assert.New(t) to all test functions
- Use require.NoError() for critical error checks
- Use assert for value checks and comparisons
- Aligns with testing patterns used in rest of package

ffly90 committed:
Fix duplicate account metrics with deduplication

- Use map to deduplicate accounts with same name
- Keep last seen value when duplicates occur
- Prevents 'collected metric with same name and label values' error
- Add test case validating deduplication behavior

abhinavDhulipala commented on contrib/prometheus-slurm-exporter:
```diff
@@ -0,0 +1 @@
+SLURM_EXPORTER_ARGS='-web.listen-address=:9341 -slurm.cli-fallback -slurm.collect-fairshare'
```
Apologies, I should've caught this in my original review, but this file needs a corresponding `contrib/prometheus-slurm-exporter.license` with the same license string. You can test this pre-commit locally by doing the following:

```bash
just init
source venv/bin/activate
pre-commit run
```

abhinavDhulipala commented on contrib/prometheus-slurm-exporter.service:
```diff
@@ -0,0 +1,11 @@
+[Unit]
```
Ditto

abhinavDhulipala reviewed (changes requested):
It's the darn reuse lint. We need special license files like the rest of the fixtures. I've added a way to iterate on these pre-commits locally so you don't have to wait for this workflow to run.

ffly90 commented:
Sorry I cannot get my supervisor to sign the Meta CLA and will therefore not be able to contribute.
